File: app/src/main/java/com/hiddenramblings/tagmo/nfctech/TagArray.java
/*
 * ====================================================================
 * Copyright (C) 2022 AbandonedCart @ TagMo
 * ====================================================================
 */

package com.hiddenramblings.tagmo.nfctech;

import android.content.Context;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.nfc.FormatException;
import android.nfc.Tag;
import android.nfc.tech.IsoDep;
import android.nfc.tech.MifareClassic;
import android.nfc.tech.MifareUltralight;
import android.nfc.tech.Ndef;
import android.nfc.tech.NdefFormatable;
import android.os.Build;

import androidx.annotation.Nullable;
import androidx.documentfile.provider.DocumentFile;

import com.hiddenramblings.tagmo.R;
import com.hiddenramblings.tagmo.TagMo;
import com.hiddenramblings.tagmo.amiibo.Amiibo;
import com.hiddenramblings.tagmo.amiibo.AmiiboFile;
import com.hiddenramblings.tagmo.amiibo.AmiiboManager;
import com.hiddenramblings.tagmo.amiibo.KeyManager;
import com.hiddenramblings.tagmo.browser.Preferences;
import com.hiddenramblings.tagmo.eightbit.io.Debug;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.BufferUnderflowException;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Locale;

public class TagArray {

    public static String getTagTechnology(Tag tag) {
        final Context context = TagMo.getContext();
        String type = context.getString(R.string.unknown_type);
        for (String tech : tag.getTechList()) {
            if (MifareClassic.class.getName().equals(tech)) {
                switch (MifareClassic.get(tag).getType()) {
                    default:
                    case MifareClassic.TYPE_CLASSIC:
                        type = context.getString(R.string.mifare_classic);
                        break;
                    case MifareClassic.TYPE_PLUS:
                        type = context.getString(R.string.mifare_plus);
                        break;
                    case MifareClassic.TYPE_PRO:
                        type = context.getString(R.string.mifare_pro);
                        break;
                }
                return type;
            } else if (MifareUltralight.class.getName().equals(tech)) {
                switch (MifareUltralight.get(tag).getType()) {
                    default:
                    case MifareUltralight.TYPE_ULTRALIGHT:
                        type = context.getString(R.string.mifare_ultralight);
                        break;
                    case MifareUltralight.TYPE_ULTRALIGHT_C:
                        type = context.getString(R.string.mifare_ultralight_c);
                        break;
                }
                return type;
            } else if (IsoDep.class.getName().equals(tech)) {
                return context.getString(R.string.isodep);
            } else if (Ndef.class.getName().equals(tech)) {
                return context.getString(R.string.ndef);
            } else if (NdefFormatable.class.getName().equals(tech)) {
                return context.getString(R.string.ndef_formatable);
            }
        }
        return type;
    }

    private static Preferences mPrefs = new Preferences(TagMo.getContext());

    public static boolean isPowerTag(NTAG215 mifare) {
        if (mPrefs.power_tag_support()) {
            byte[] signature = mifare.transceive(NfcByte.POWERTAG_SIG);
            return null != signature && TagArray.compareRange(signature,
                    NfcByte.POWERTAG_SIGNATURE, NfcByte.POWERTAG_SIGNATURE.length);
        }
        return false;
    }

    public static boolean isElite(NTAG215 mifare) {
        if (mPrefs.elite_support()) {
            byte[] signature = mifare.readSignature(false);
            byte[] page10 = TagArray.hexToByteArray("FFFFFFFFFF");
            return null != signature && TagArray.compareRange(signature, page10,
                    32 - page10.length, signature.length);
        }
        return false;
    }

    static boolean compareRange(byte[] data, byte[] data2, int offset, int len) {
        for (int i = 0, j = offset; j < len; i++, j++) {
            if (data[j] != data2[i])
                return false;
        }
        return true;
    }

    public static boolean compareRange(byte[] data, byte[] data2, int len) {
        return compareRange(data, data2, 0, len);
    }

    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02X", b));
        }
        return sb.toString();
    }

    public static byte[] hexToByteArray(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i + 1), 16));
        }
        return data;
    }

    public static byte[] longToBytes(long x) {
        ByteBuffer buffer = ByteBuffer.allocate(
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.N ? Long.BYTES : 8
        );
        buffer.putLong(x);
        return buffer.array();
    }

    public static long bytesToLong(byte[] bytes) {
        ByteBuffer buffer = ByteBuffer.allocate(
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.N ? Long.BYTES : 8
        );
        buffer.put(bytes);
        buffer.flip(); // need flip
        try {
            return buffer.getLong();
        } catch (BufferUnderflowException bue) {
            return (long) buffer.getInt();
        }
    }

    public static long hexToLong(String s) {
        long result = 0;
        for (int i = 0; i < s.length(); i++) {
            result = (result << 4) + ((long) Character.digit(s.charAt(i), 16));
        }
        return result;
    }

    static String hexToString(String hex) {
        StringBuilder output = new StringBuilder();
        for (int i = 0; i < hex.length(); i += 2) {
            output.append((char) Integer.parseInt(hex.substring(i, i + 2), 16));
        }
        return output.toString();
    }

    @Nullable
    public static void validateData(byte[] data) throws Exception {
        final Context context = TagMo.getContext();
        if (null == data)
            throw new IOException(context.getString(R.string.invalid_data_null));
        /* TagWriter.splitPages(data) */
        if (data.length == NfcByte.KEY_FILE_SIZE || data.length == NfcByte.KEY_FILE_SIZE * 2)
            throw new IOException(context.getString(R.string.invalid_tag_key));
        else if (data.length < NfcByte.TAG_DATA_SIZE)
            throw new IOException(context.getString(
                    R.string.invalid_data_size, data.length, NfcByte.TAG_DATA_SIZE
            ));

        byte[][] pages = new byte[data.length / NfcByte.PAGE_SIZE][];
        for (int i = 0, j = 0; i < data.length; i += NfcByte.PAGE_SIZE, j++) {
            pages[j] = Arrays.copyOfRange(data, i, i + NfcByte.PAGE_SIZE);
        }

        if (pages[0][0] != (byte) 0x04)
            throw new Exception(context.getString(R.string.invalid_tag_prefix));

        if (pages[2][2] != (byte) 0x0F || pages[2][3] != (byte) 0xE0)
            throw new Exception(context.getString(R.string.invalid_tag_lock));

        if (pages[3][0] != (byte) 0xF1 || pages[3][1] != (byte) 0x10
                || pages[3][2] != (byte) 0xFF || pages[3][3] != (byte) 0xEE)
            throw new Exception(context.getString(R.string.invalid_tag_cc));

        if (pages[0x82][0] != (byte) 0x01 || pages[0x82][1] != (byte) 0x0 || pages[0x82][2] != (byte) 0x0F)
            throw new Exception(context.getString(R.string.invalid_tag_dynamic));

        if (pages[0x83][0] != (byte) 0x0 || pages[0x83][1] != (byte) 0x0
                || pages[0x83][2] != (byte) 0x0 || pages[0x83][3] != (byte) 0x04)
            throw new Exception(context.getString(R.string.invalid_tag_cfg_zero));

        if (pages[0x84][0] != (byte) 0x5F || pages[0x84][1] != (byte) 0x0
                || pages[0x84][2] != (byte) 0x0 || pages[0x84][3] != (byte) 0x00)
            throw new Exception(context.getString(R.string.invalid_tag_cfg_one));
    }

    public static void validateNtag(NTAG215 mifare, byte[] tagData, boolean validateNtag)
            throws Exception {
        final Context context = TagMo.getContext();
        if (null == tagData) throw new IOException(context.getString(R.string.no_source_data));

        if (validateNtag) {
            try {
                byte[] versionInfo = mifare.transceive(new byte[]{(byte) 0x60});
                if (null == versionInfo  || versionInfo.length != 8)
                    throw new Exception(context.getString(R.string.error_tag_version));
                if (versionInfo[0x02] != (byte) 0x04 || versionInfo[0x06] != (byte) 0x11)
                    throw new FormatException(context.getString(R.string.error_tag_specs));
            } catch (Exception e) {
                Debug.Warn(R.string.error_version, e);
                throw e;
            }
        }

        byte[] pages = mifare.readPages(0);
        if (null == pages  || pages.length != NfcByte.PAGE_SIZE * 4)
            throw new Exception(context.getString(R.string.fail_read_size));

        if (!TagArray.compareRange(pages, tagData, 9))
            throw new Exception(context.getString(R.string.fail_mismatch_uid));

        Debug.Info(TagWriter.class, R.string.validation_success);
    }

    public static String decipherFilename(AmiiboManager amiiboManager, byte[] tagData, boolean verified) {
        String status = "";
        if (verified) {
            try {
                validateData(tagData);
                status = "Validated";
            } catch (Exception e) {
                Debug.Warn(e);
                status = e.getMessage();
            }
        }
        try {
            long amiiboId = Amiibo.dataToId(tagData);
            String name = Amiibo.idToHex(amiiboId);
            if (null != amiiboManager) {
                Amiibo amiibo = amiiboManager.amiibos.get(amiiboId);
                if (null != amiibo && null != amiibo.name) {
                    name = amiibo.name.replace(File.separatorChar, '-');
                }
            }
            
            String uidHex = bytesToHex(Arrays.copyOfRange(tagData, 0, 9));
            if (verified)
                return String.format(Locale.ROOT, "%1$s[%2$s]-%3$s.bin", name, uidHex, status);
            else
                return String.format(Locale.ROOT, "%1$s[%2$s].bin", name, uidHex);
        } catch (Exception ex) {
            Debug.Warn(ex);
        }
        return "";
    }

    public static byte[] getValidatedData(KeyManager keyManager, byte[] data) throws Exception {
        if (null == data ) return null;
        try {
            TagArray.validateData(data);
            data = keyManager.decrypt(data);
        } catch (Exception e) {
            data = keyManager.encrypt(data);
            TagArray.validateData(data);
            data = keyManager.decrypt(data);
        }
        return keyManager.encrypt(data);
    }

    public static byte[] getValidatedFile(KeyManager keyManager, File file) throws Exception {
        return getValidatedData(keyManager, TagReader.readTagFile(file));
    }

    public static byte[] getValidatedDocument(
            KeyManager keyManager, Uri fileUri) throws Exception {
        return getValidatedData(keyManager, TagReader.readTagDocument(fileUri));
    }

    public static byte[] getValidatedDocument(
            KeyManager keyManager, DocumentFile file) throws Exception {
        return getValidatedData(keyManager, TagReader.readTagDocument(file.getUri()));
    }

    public static byte[] getValidatedData(KeyManager keyManager, AmiiboFile file) throws Exception {
        return null != file.getData() ? file.getData() : null != file.getDocUri()
                ? TagArray.getValidatedDocument(keyManager, file.getDocUri())
                : TagArray.getValidatedFile(keyManager, file.getFilePath());
    }

    public static String writeBytesToFile(File directory, String name, byte[] tagData)
            throws IOException {
        File binFile = new File(directory, name);
        new FileOutputStream(binFile).write(tagData);
        try {
            MediaScannerConnection.scanFile(TagMo.getContext(),
                    new String[] { binFile.getAbsolutePath() },
                    null, null);
        } catch (Exception e) {
            Debug.Info(e);
        }
        return binFile.getAbsolutePath();
    }

    public static String writeBytesToDocument(
            Context context, DocumentFile directory, String name, byte[] tagData
    ) throws IOException {
        DocumentFile newFile = directory.createFile(context.getResources()
                .getStringArray(R.array.mimetype_bin)[0], name + ".bin");
        if (null != newFile) {
            context.getContentResolver().openOutputStream(newFile.getUri()).write(tagData);
        }
        return null != newFile ? newFile.getName() : null;
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/nfctech/TagArray.kt
/*
 * ====================================================================
 * Copyright (C) 2022 AbandonedCart @ TagMo
 * ====================================================================
 */

package com.hiddenramblings.tagmo.nfctech

import android.content.Context
import android.media.MediaScannerConnection
import android.net.Uri
import android.nfc.FormatException
import android.nfc.Tag
import android.nfc.tech.*
import android.text.Editable
import androidx.documentfile.provider.DocumentFile
import com.hiddenramblings.tagmo.Preferences
import com.hiddenramblings.tagmo.R
import com.hiddenramblings.tagmo.TagMo
import com.hiddenramblings.tagmo.amiibo.Amiibo
import com.hiddenramblings.tagmo.amiibo.AmiiboFile
import com.hiddenramblings.tagmo.amiibo.AmiiboManager
import com.hiddenramblings.tagmo.amiibo.KeyManager
import com.hiddenramblings.tagmo.eightbit.io.Debug
import com.hiddenramblings.tagmo.eightbit.os.Storage
import com.hiddenramblings.tagmo.eightbit.os.Version
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.BufferUnderflowException
import java.nio.ByteBuffer
import java.util.*


object TagArray {
    @JvmStatic
    fun Tag?.technology(): String {
        with (TagMo.appContext) {
            var type = getString(R.string.unknown_type)
            if (null == this@technology) return type
            techList.forEach {
                when {
                    MifareClassic::class.java.name == it -> {
                        type = when (MifareClassic.get(this@technology).type) {
                            MifareClassic.TYPE_CLASSIC -> getString(R.string.mifare_classic)
                            MifareClassic.TYPE_PLUS -> getString(R.string.mifare_plus)
                            MifareClassic.TYPE_PRO -> getString(R.string.mifare_pro)
                            else -> getString(R.string.mifare_classic)
                        }
                        return type
                    }
                    MifareUltralight::class.java.name == it -> {
                        type = when (MifareUltralight.get(this@technology).type) {
                            MifareUltralight.TYPE_ULTRALIGHT -> getString(R.string.mifare_ultralight)
                            MifareUltralight.TYPE_ULTRALIGHT_C -> getString(R.string.mifare_ultralight_c)
                            else -> getString(R.string.mifare_ultralight)
                        }
                        return type
                    }
                    IsoDep::class.java.name == it -> { return getString(R.string.isodep) }
                    Ndef::class.java.name == it -> { return getString(R.string.ndef) }
                    NdefFormatable::class.java.name == it -> { return getString(R.string.ndef_formatable) }
                }
            }
            return type
        }
    }

    private val mPrefs = Preferences(TagMo.appContext)
    @JvmStatic
    fun isPowerTag(mifare: NTAG215?): Boolean {
        if (mPrefs.powerTagEnabled()) {
            return mifare?.transceive(NfcByte.POWERTAG_SIG)?.let {
                compareRange(it, NfcByte.POWERTAG_SIGNATURE, NfcByte.POWERTAG_SIGNATURE.size)
            } ?: false
        }
        return false
    }

    @JvmStatic
    fun isElite(mifare: NTAG215?): Boolean {
        if (mPrefs.eliteEnabled()) {
            val signature = mifare?.readSignature(false)
            val page10 = hexToByteArray("FFFFFFFFFF")
            return signature?.let {
                compareRange(it, page10, 32 - page10.size, it.size)
            } ?: false
        }
        return false
    }

    private fun compareRange(data: ByteArray, data2: ByteArray?, offset: Int, len: Int): Boolean {
        var i = 0
        var j = offset
        while (j < len) {
            if (data2?.get(i) != data[j]) return false
            i++
            j++
        }
        return true
    }

    @JvmStatic
    fun compareRange(data: ByteArray, data2: ByteArray?, len: Int): Boolean {
        return compareRange(data, data2, 0, len)
    }

    @JvmStatic
    fun bytesToHex(bytes: ByteArray?): String {
        val sb = StringBuilder()
        bytes?.forEach { sb.append(String.format("%02X", it)) }
        return sb.toString()
    }

    @JvmStatic
    fun hexToByteArray(s: String): ByteArray {
        val len: Int = s.length
        val data = ByteArray(len / 2)
        var i = 0
        while (i < len) {
            data[i / 2] = ((Character.digit(s[i], 16) shl 4)
                    + Character.digit(s[i + 1], 16)).toByte()
            i += 2
        }
        return data
    }

    fun longToBytes(x: Long): ByteArray {
        return ByteBuffer.allocate(
            if (Version.isNougat) java.lang.Long.BYTES else 8
        ).putLong(x).array()
    }

    fun bytesToLong(bytes: ByteArray): Long {
        val buffer = ByteBuffer.allocate(
            if (Version.isNougat) java.lang.Long.BYTES else 8
        ).put(bytes).also { it.flip() }
        return try {
            buffer.long
        } catch (bue: BufferUnderflowException) {
            try {
                buffer.int.toLong()
            } catch (bue: BufferUnderflowException) {
                buffer.short.toLong()
            }
        }
    }

    fun hexToLong(hex: String): Long {
        var result: Long = 0
        try {
            result = hex.toLong(16)
        } catch (nf: NumberFormatException) {
            hex.forEach { result = (result shl 4) + Character.digit(it, 16).toLong() }
        }
        return result
    }

    fun hexToString(hex: String): String {
        val output = StringBuilder()
        var i = 0
        while (i < hex.length) {
            try {
                output.append(Integer.parseInt(hex.substring(i, i + 2), 16).toChar())
            } catch (nf: NumberFormatException) {
                Debug.warn(nf)
                output.clear()
                break
            }
            i += 2
        }
        return output.toString()
    }

    fun bytesToString(bytes: ByteArray?): String {
        return hexToString(bytesToHex(bytes))
    }

    @JvmStatic
    @Throws(Exception::class)
    fun validateData(data: ByteArray?) {
        with (TagMo.appContext) {
            if (null == data) throw IOException(getString(R.string.invalid_data_null))
            /* TagWriter.splitPages(data) */
            if (data.size == NfcByte.KEY_FILE_SIZE || data.size == NfcByte.KEY_RETAIL_SZ)
                throw IOException(getString(R.string.invalid_tag_key))
            else if (data.size < NfcByte.TAG_DATA_SIZE)
                throw IOException(
                    getString(R.string.invalid_data_size, data.size, NfcByte.TAG_DATA_SIZE)
                )
            val pages = arrayOfNulls<ByteArray>(data.size / NfcByte.PAGE_SIZE)
            var i = 0
            var j = 0
            while (i < data.size) {
                pages[j] = Arrays.copyOfRange(data, i, i + NfcByte.PAGE_SIZE)
                i += NfcByte.PAGE_SIZE
                j++
            }
            when {
                pages[0]?.let {
                    it[0] != 0x04.toByte()
                } ?: true -> throw Exception(getString(R.string.invalid_tag_prefix))

                pages[2]?.let {
                    it[2] != 0x0F.toByte() || it[3] != 0xE0.toByte()
                } ?: true -> throw Exception(getString(R.string.invalid_tag_lock))

                pages[3]?.let {
                    it[0] != 0xF1.toByte() || it[1] != 0x10.toByte()
                            || it[2] != 0xFF.toByte() || it[3] != 0xEE.toByte()
                } ?: true -> throw Exception(getString(R.string.invalid_tag_cc))

                pages[0x82]?.let {
                    it[0] != 0x01.toByte() || it[1] != 0x0.toByte() || it[2] != 0x0F.toByte()
                } ?: true -> throw Exception(getString(R.string.invalid_tag_dynamic))

                pages[0x83]?.let {
                    it[0] != 0x0.toByte() || it[1] != 0x0.toByte()
                            || it[2] != 0x0.toByte() || it[3] != 0x04.toByte()
                } ?: true -> throw Exception(getString(R.string.invalid_tag_cfg_zero))

                pages[0x84]?.let {
                    it[0] != 0x5F.toByte() || it[1] != 0x0.toByte()
                            || it[2] != 0x0.toByte() || it[3] != 0x00.toByte()
                } ?: true -> throw Exception(getString(R.string.invalid_tag_cfg_one))

                else -> {}
            }
        }
    }

    @JvmStatic
    @Throws(Exception::class)
    fun validateNtag(mifare: NTAG215, tagData: ByteArray?, validateNtag: Boolean) {
        with (TagMo.appContext) {
            if (null == tagData) throw IOException(getString(R.string.no_source_data))
            if (validateNtag) {
                try {
                    mifare.transceive(byteArrayOf(0x60.toByte()))?.let {
                        if (it.size != 8) throw Exception(getString(R.string.error_tag_version))
                        if (it[0x02] != 0x04.toByte() || it[0x06] != 0x11.toByte())
                            throw FormatException(getString(R.string.error_tag_specs))
                    } ?: throw Exception(getString(R.string.error_tag_version))
                } catch (e: Exception) {
                    Debug.warn(R.string.error_version, e)
                    throw e
                }
            }
            val pages = mifare.readPages(0)
            if (null == pages || pages.size != NfcByte.PAGE_SIZE * 4)
                throw Exception(getString(R.string.fail_read_size))
            if (!compareRange(pages, tagData, 9))
                throw Exception(getString(R.string.fail_mismatch_uid))
            Debug.info(TagWriter::class.java, R.string.validation_success)
        }
    }

    @JvmStatic
    fun decipherFilename(amiibo: Amiibo, tagData: ByteArray?, verified: Boolean): String {
        var status = ""
        if (verified) {
            status = try {
                validateData(tagData)
                "Validated"
            } catch (e: Exception) {
                Debug.warn(e)
                Debug.getExceptionClass(e)
            }
        }
        try {
            val name = amiibo.name?.replace(File.separatorChar, '-')
            val uidHex = bytesToHex(tagData?.copyOfRange(0, 9))
            return if (verified) String.format(
                Locale.ROOT, "%1\$s[%2\$s]-%3\$s.bin", name, uidHex, status
            ) else String.format(
                Locale.ROOT, "%1\$s[%2\$s].bin", name, uidHex
            )
        } catch (ex: Exception) {
            Debug.warn(ex)
        }
        return ""
    }

    @JvmStatic
    fun decipherFilename(
        amiiboManager: AmiiboManager?, tagData: ByteArray?, verified: Boolean
    ): String {
        try {
            amiiboManager?.let {
                return it.amiibos[Amiibo.dataToId(tagData)]?.let { amiibo ->
                    decipherFilename(amiibo, tagData, verified)
                } ?: ""
            }
        } catch (ex: Exception) {
            Debug.warn(ex)
        }
        return ""
    }

    @JvmStatic
    @Throws(Exception::class)
    fun getValidatedData(keyManager: KeyManager?, data: ByteArray?): ByteArray? {
        if (null == keyManager || null == data) return null
        var validated = data
        try {
            validateData(validated)
        } catch (e: Exception) {
            validated = keyManager.encrypt(validated)
            validateData(validated)
        }
        validated = keyManager.decrypt(validated)
        return keyManager.encrypt(validated)
    }

    @JvmStatic
    @Throws(Exception::class)
    fun getValidatedFile(keyManager: KeyManager?, file: File?): ByteArray? {
        return getValidatedData(keyManager, TagReader.readTagFile(file))
    }

    @Throws(Exception::class)
    fun getValidatedDocument(keyManager: KeyManager?, fileUri: Uri): ByteArray? {
        return getValidatedData(keyManager, TagReader.readTagDocument(fileUri))
    }

    @Throws(Exception::class)
    fun getValidatedDocument(
        keyManager: KeyManager?, file: DocumentFile
    ): ByteArray? {
        return getValidatedData(keyManager, TagReader.readTagDocument(file.uri))
    }

    @JvmStatic
    @Throws(Exception::class)
    fun getValidatedData(keyManager: KeyManager?, file: AmiiboFile): ByteArray? {
        return file.data ?: file.docUri?.let { getValidatedDocument(keyManager, it) }
        ?: file.filePath?.let { getValidatedFile(keyManager, it) }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeBytesToFile(directory: File?, name: String, tagData: ByteArray?): String {
        val binFile = File(directory, name)
        FileOutputStream(binFile).use { it.write(tagData) }
        try {
            MediaScannerConnection.scanFile(
                TagMo.appContext, arrayOf(binFile.absolutePath), null, null
            )
        } catch (e: Exception) { Debug.info(e) }
        return binFile.absolutePath
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeBytesToDocument(
        context: Context, directory: DocumentFile, name: String, tagData: ByteArray?
    ): String? {
        val newFile = directory.createFile(
            context.resources.getStringArray(R.array.mimetype_bin)[0], name
        )
        newFile?.let { file ->
            context.contentResolver.openOutputStream(file.uri).use { it?.write(tagData) }
        }
        return newFile?.name
    }

    fun writeBytesWithName(context: Context, fileName: String?, tagData: ByteArray?) : String? {
        return with (Preferences(context.applicationContext)) {
            fileName?.let { name ->
                if (isDocumentStorage) {
                    val rootDocument = browserRootDocument()?.let { uri ->
                        DocumentFile.fromTreeUri(context, Uri.parse(uri))
                    } ?: throw NullPointerException()
                    writeBytesToDocument(context, rootDocument, name, tagData)
                } else {
                    writeBytesToFile(
                        Storage.getDownloadDir("TagMo", "Backups"), name, tagData
                    )
                }
            }
        }
    }

    fun writeBytesWithName(context: Context, input: Editable?, tagData: ByteArray?) : String? {
        return writeBytesWithName(context, input?.toString(), tagData)
    }
}